flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java

```diff
@@ -131,46 +137,119 @@ private static ReporterSetup createReporterSetup(String reporterName, MetricConf
 		}
 		return reporterArguments;
 	}
+
+	private static Collection<MetricReporterFactory> loadReporterFactories() {
+		final ServiceLoader<MetricReporterFactory> serviceLoader = ServiceLoader.load(MetricReporterFactory.class);
+
+		final Collection<MetricReporterFactory> reporterFactories = new ArrayList<>();
+		final Iterator<MetricReporterFactory> factoryIterator = serviceLoader.iterator();
+		// do not use streams or for-each loops here because they do not allow catching individual ServiceConfigurationErrors
+		// such an error might be caused if the META-INF/services contains an entry to a non-existing factory class
+		while (factoryIterator.hasNext()) {
+			try {
+				MetricReporterFactory factory = factoryIterator.next();
+				reporterFactories.add(factory);
+			} catch (Exception | ServiceConfigurationError e) {
+				LOG.warn("Error while loading reporter factory.", e);
+			}
+		}
+
+		return Collections.unmodifiableCollection(reporterFactories);
+	}
+
+	private static Optional<MetricReporter> loadReporter(
+			final String reporterName,
+			final Configuration reporterConfig,
+			final Collection<MetricReporterFactory> reporterFactories)
+			throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+
+		final String reporterClassName = reporterConfig.getString(ConfigConstants.METRICS_REPORTER_CLASS_SUFFIX, null);
+		final String factoryClassName = reporterConfig.getString(ConfigConstants.METRICS_REPORTER_FACTORY_CLASS_SUFFIX, null);
+
+		if (factoryClassName != null) {
+			return loadViaFactory(factoryClassName, reporterName, reporterConfig, reporterFactories);
+		}
+
+		if (reporterClassName != null) {
+			final Class<?> reporterClass = Class.forName(reporterClassName);
+
+			final InstantiateViaFactory alternativeFactoryAnnotation = reporterClass.getAnnotation(InstantiateViaFactory.class);
+			if (alternativeFactoryAnnotation != null) {
+				final String alternativeFactoryClassName = alternativeFactoryAnnotation.factoryClassName();
+				LOG.info("The reporter configuration of {} is out-dated (but still supported)." +
+						" Please configure a factory class instead: '{}{}.{}: {}' to ensure that the configuration" +
+						" continues to work with future versions.",
+					reporterName,
+					ConfigConstants.METRICS_REPORTER_PREFIX,
+					reporterName,
+					ConfigConstants.METRICS_REPORTER_FACTORY_CLASS_SUFFIX,
+					alternativeFactoryClassName);
+				return loadViaFactory(alternativeFactoryClassName, reporterName, reporterConfig, reporterFactories);
+			}
+
+			return Optional.of((MetricReporter) reporterClass.newInstance());
+		}
+
+		LOG.warn("No reporter class nor factory set for reporter {}. Metrics might not be exposed/reported.", reporterName);
+		return Optional.empty();
+	}
+
+	private static Optional<MetricReporter> loadViaFactory(
+			final String factoryClassName,
+			final String reporterName,
+			final Configuration reporterConfig,
+			final Collection<MetricReporterFactory> reporterFactories) {
```

Comment:
I think it would make sense to construct a `Map<String, MetricReporterFactory>` for easier lookups at an earlier place.